### DISCUSSION TITLE
GTM setup

### DISCUSSION THREAD

**Discussion by oxodesign** [vercel/next.js#15518] [Help] [answered]
Hi all

We have GA example (https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics) and based on that I tried to implement GTM but without any luck.

Is it possible to provide us with an working example on how to implement GTM in Next.js? Or if someone has a snippet they want to share with us?

All help appreciated

**Comment by jamesmosier** [comment]
What about the example did not work? 

Here is another example I use in an application:

```jsx
import Document, { Html, Head, Main, NextScript } from 'next/document';

export default class CustomDocument extends Document {
  renderGTMSnippet() {
    return (
      <script
        dangerouslySetInnerHTML={{
          __html: `
        (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
        new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
        j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
        'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
        })(window,document,'script','dataLayer', '${process.env.GTM_KEY}');
        `,
        }}
      />
    );
  }

  render() {
    return (
      <Html lang="en">
        <Head>
          {this.renderGTMSnippet()}
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

```

**Reply by oxodesign** [reply]
@jamesmosier This does not track page view (page changes).

**Reply by benjaminhoffman** [reply]
@oxodesign That's because of how react / SPA work, and not related to nextjs per se. If you'd like to fire GTM correctly on each page navigation, be sure to enable History change trigger, as described here: https://support.google.com/tagmanager/answer/7679322?hl=en

Then attach the trigger to whatever tag you'd like fired on pageviews, like Google Analytics. See here for my setup: 
![gtm](https://user-images.githubusercontent.com/6520022/137026385-47879af2-f4ce-4586-bee3-e069ac89b63b.png)

EDIT: I see this was already answered below 🤷🏽‍♂️

**Comment by mahmoudalwadia** [comment]
I think you can use  [react-gtm-module](https://www.npmjs.com/package/react-gtm-module) we are using it in our project and it's work fine, you need to make the setup in your `_app` then you can use it anywhere.

**Comment by good-idea** [comment, marked as answer]
This is not a next thing so much as a PWA issue. A typical "Page view" trigger in GTM will only fire when the browser first loads. You'll need to add a History Change trigger to your tag:

![image](https://user-images.githubusercontent.com/11514928/88579377-c90df780-cffe-11ea-8867-24ba62dc7d0a.png)

I didn't realize that my app had the same issue. I just added the history change trigger and it's working:

First page:

![image](https://user-images.githubusercontent.com/11514928/88579455-e773f300-cffe-11ea-9dc7-3faa1a420f3e.png)

Navigating to a new page:

![image](https://user-images.githubusercontent.com/11514928/88579485-f65aa580-cffe-11ea-9743-5ab0a9232d9a.png)

Note the "Page view" tag in the GTM console - `Fired 2 time(s)`

**Reply by oxodesign** [reply]
@good-idea The history trigger was mentioned another place as well! Thanks for sharing your settings.

**Reply by seonghyeonkimm** [reply]
@oxodesign I followed your setting. But I found out when route.back() is called, gtm caught history change twice. Did you face this issue too? 

Edit:
I found out that I can skip trigger the tag adding condition to skip popstate history change.

**Comment by oxodesign** [comment]
I also blogged on how to fire this manually on Next.js, here is the blog post:
https://dev.to/ornio/add-google-analytics-through-gtm-google-tag-manager-on-next-js-5e4f

**Reply by roelvan** [reply]
broken link
https://web.archive.org/web/20201121045651/https://dev.to/ornio/add-google-analytics-through-gtm-google-tag-manager-on-next-js-5e4f

**Reply by oxodesign** [reply]
It's working fin for me!

**Reply by roelvan** [reply]
Found the culprit: adblocker! :-)

**Reply by oxodesign** [reply]
> Found the culprit: adblocker! :-)

Ehhh...! Good you found the problem!

**Comment by ribeirogab** [comment]
The easiest way is to use a `react-gtm-module` library.
A code example from the **_app** file:
```ts
import { FC, useEffect } from 'react';
import { AppProps } from 'next/app';
import TagManager from 'react-gtm-module';

const MyApp: FC<AppProps> = ({ Component, pageProps }) => {
  // Google Tag Manager
  useEffect(() => {
    TagManager.initialize({ gtmId: 'GTM-XXXXXX' });
  }, []);

  return <Component {...pageProps} />;
};

export default MyApp;
```

**Reply by cniska** [reply]
@ribeirogab Is there any specific reason why you call `TagManager.initialize` from a side-effect? 

According to the [readme](https://github.com/alinemorelli/react-gtm/blob/master/README.md) you can simply call `initialize` outside of the component tree:

**_app.ts**
```typescript
import React from 'react';
import {AppProps} from 'next/app';
import TagManager from 'react-gtm-module';

TagManager.initialize({ gtmId: 'GTM-XXXXXX' });

function MyApp({Component, pageProps}: AppProps) {
  return <Component {...pageProps} />;
}

export default MyApp;
```

**Reply by zhicong** [reply]
Just happened to implement this in my Next.js app myself - `react-gtm-module` expects to be called only on the _client_ side, so `useEffect` ensures it is not called during SSR. Otherwise, you'll get `ReferenceError: document is not defined`.

EDIT: @selrond Thanks for pointing out the typo!

**Reply by selrond** [reply]
@zhicong 
> `react-gtm-module` expects to be called only on the server side

Did you mean client-side there?

**Reply by rodilo** [reply]
Can it be an issue that sometimes GTM is loaded twice in the app?
You can see this when you log `dataLayer` object, there will be two `gtm.start` events.

![image](https://user-images.githubusercontent.com/32063803/146147178-214965da-91e3-4825-9f1e-fc036833d0bc.png)

**Comment by benjaminhoffman** [comment]
What's strange to me is how segment & gtm are becoming the de facto standard for analytics pageview & event tracking, yet there doesn't seem to be a clear out of the box solution for us.  For example, see `routeChangeComplete` sits in `Page.js` instead of `_app.js` https://github.com/vercel/next.js/blob/canary/examples/with-segment-analytics/components/Page.js#L6

I'm still working my through this so open to feedback but I think the better solution is to just use GTM's History Change trigger and _not_ the `routeChangeComplete` router event.

**Reply by oxodesign** [reply]
@benjaminhoffman with new GTM>GA4 and History Change is working perfect for me :)

**Reply by benjaminhoffman** [reply]
@oxodesign I know, I know... but GA4 is so... lame lol

**Reply by oxodesign** [reply]
> @oxodesign I know, I know... but GA4 is so... lame lol

I found it much better regarding events and funnels :)

**Comment by morganfeeney** [comment]
I found this discussion really helpful. I discovered a while back that using `routeChangeComplete` is ok for a simple page view where you want to send a url to GTM, but not a lot else.
After going through the motions of dealing with Next.js and GTM in my day jobs I've distilled some practical working knowledge here: https://morganfeeney.com/how-to/integrate-google-tag-manager-with-next-js

**Reply by JinMABrett** [reply]
Hi Morgan, I followed your blog post on getting everything set up, but I have some questions (sorry i'm very new to programming AND GTM). After setup, all I seem to get is a duplicate "History" tag. It's marked as "VirtualPageView" but has the exact same data as the "history" change above it that was already there before implementing it the way you laid out. 

Has GTM since your post updated and is working fine out the box and your method is no longer necessary? My site ended up being mostly static pages so maybe that's why it seems unnecessary? Or are there other benefits that it's giving me that I'm not properly understanding even if GTM is recognizing the history change without it.

**Reply by morganfeeney** [reply]
Really hard to say based on words alone.

**Reply by oxodesign** [reply]
> Hi Morgan, I followed your blog post on getting everything set up, but I have some questions (sorry i'm very new to programming AND GTM). After setup, all I seem to get is a duplicate "History" tag. It's marked as "VirtualPageView" but has the exact same data as the "history" change above it that was already there before implementing it the way you laid out.
> 
> Has GTM since your post updated and is working fine out the box and your method is no longer necessary? My site ended up being mostly static pages so maybe that's why it seems unnecessary? Or are there other benefits that it's giving me that I'm not properly understanding even if GTM is recognizing the history change without it.

@JinMABrett are you doning any page transition/routing at the same time?

**Reply by JinMABrett** [reply]
I don't know if it's necessary, but I am using next/link if that's what you mean (sorry still very new to all this) and it has given me very strange behavior as far as tracking my link clicks. But strangely even though it looks very strange odd behavior in debug view, when I view the live data in GA4 it seems to be tracking my page visits just fine. 

(some strange behavior is, next/link links don't seem to populate any data, such as "click text" or "click ID"  but if I use a material UI button and just use the href inside the button without using next/link it recognizes the page transitions in debug view as well as passes all proper data just fine. But if I click a next/link link after that, it populates all the data with whatever matches the last non next/link link i clicked was. I'm overall just very confused )

**Comment by jaxramus** [comment]
Hi friends, here is how to set up GTM in nextjs.

Install `react-gtm-module`:

```console
npm install react-gtm-module --save
```

Go into your `pages/_app.js` file, and initialize the GTM module inside `useEffect`:

```js
import '../styles/globals.css'
import { useEffect } from 'react'
import TagManager from 'react-gtm-module'

function MyApp({ Component, pageProps }) {
  useEffect(() => {
    TagManager.initialize({ gtmId: "GTM-XXXXXXX" })
  }, [])

  return <Component {...pageProps} />
}

export default MyApp
```

Be sure to replace `GTM-XXXXXXX` with your full GTM ID.

edit: looks like @ribeirogab beat me to this, my bad.

**Reply by jpcmf** [reply]
Is it possible to use two GTM IDs? One for the main pages and another for the blog/post pages in the same Next.js application? Does this make sense? I have been trying to handle it without success.

**Comment by garrettboatman** [comment]
For simplicity in Next 13, the best solution for me was using `next/script` in `app/layout.tsx`

```jsx
/* app/layout.tsx */

import Script from 'next/script';

const RootLayout = async ({ children }: { children: React.ReactNode }) => {
  return (
    <html lang="en">
      <head>
        {/* Google Tag Manager */}
        {process.env.VERCEL_ENV === 'production' && (
          <Script id="google-tag-manager">
            {`
              (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
              new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
              j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
              'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
              })(window,document,'script','dataLayer','GTM-XXXXXXXX');
            `}
          </Script>
        )}
        {/* END Google Tag Manager */}
      </head>
      <body>
        {/* Google Tag Manager */}
        {process.env.VERCEL_ENV === 'production' && (
          <noscript>
            <iframe
              src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXXX"
              height="0"
              width="0"
              style={{ display: 'none', visibility: 'hidden' }}
            ></iframe>
          </noscript>
        )}
        {/* END Google Tag Manager */}
        {children}
      </body>
    </html>
  );
};

export default RootLayout;
```
